checks/branch_protection.go

```diff
@@ -176,20 +386,33 @@ func checkReleaseAndDevBranchProtection(
 		// so it does not provide any guarantees.
 		if branch.Protected != nil && !*branch.Protected {
 			dl.Warn("branch protection not enabled for branch '%s'", b)
-			return checker.CreateMinScoreResult(CheckBranchProtection,
-				fmt.Sprintf("branch protection not enabled on development/release branch: %s", b))
+			scores = append(scores, score)
+			continue
 		}
 
 		// The branch is protected. Check the protection.
-		score := isBranchProtected(&branch.BranchProtectionRule, b, dl)
+		score.scores.basic, score.maxes.basic = basicNonAdminProtection(&branch.BranchProtectionRule, b, dl)
+		score.scores.adminBasic, score.maxes.adminBasic = adminBasicProtection(&branch.BranchProtectionRule, b, dl)
+		score.scores.review, score.maxes.review = nonAdminReviewProtection(&branch.BranchProtectionRule)
+		score.scores.adminReview, score.maxes.adminReview = adminReviewProtection(&branch.BranchProtectionRule, b, dl)
+		score.scores.context, score.maxes.context = nonAdminContextProtection(&branch.BranchProtectionRule, b, dl)
+		score.scores.thoroughReview, score.maxes.thoroughReview =
+			nonAdminThoroughReviewProtection(&branch.BranchProtectionRule, b, dl)
+		score.scores.adminThoroughReview, score.maxes.adminThoroughReview =
+			adminThoroughReviewProtection(&branch.BranchProtectionRule, b, dl) // Do we want this?
+
```

Comment:
How about we write it this way:

```
tier1Score, tier1Passing := calculateTier1Score(branch.BranchProtectionRule)

if tier1Passing {
  tier2Score, tier2Passing := calculateTier2Score(...)
}
if tier2Passing {
  ...
}
score := tier1Score + tier2Score + ...
```

wherein, the `calculateTierXScore` returns true either if:
(i) all requirements for the tier are met, or
(ii) some fields are `nil`, indicating no data, so we don't penalize the repo.

This way we avoid the complex structs and also avoid leaking `admin token info` here in the check. Wdyt?

Reply:
We can do the tier calculation in this part, because it's iterating over all branches: we need the data from all branches to calculate the tiers. This function is collecting the result. The tier calculation is done the `computeScore()` function. Can you check it?

Reply:
> (ii) some fields are nil, indicating no data, so we don't penalize the repo.

I think without overhauling and separating out score/check (maybe not in this PR, i'm happy to give this a tackle on monday), you will need to propogate a max score since nil will either mean nil or no admin token. so i'm not sure this will work unless you get that info, in which case i think it's pretty much equivalent to what's currently implemented

Reply:
> We cannot do the tier calculation in this part, because it's iterating over all branches: we need the data from all branches to calculate the tiers. This function is collecting the result. The tier calculation is done in the computeScore() function and essentially does what you describe here. Can you check it?

Ah, I see. Then, let's calculate per-branch tier scores here and aggregate them later? So each branch will have:

```
{
  tier1: <score, passing>,
  tier2: <score, passing>,
 ...
}
```

which will be later aggregated across all branches. Would that work?

Reply:
> nil will either mean nil or no admin token

Right. And there is no way for us differentiate between the 2. Even with the knowledge that some fields are `admin-only`, a `nil` does not mean that it is because a lack of admin privileges. It simply means there was no data. So I see no reason to make speculations about why certain fields were returned `nil`.

Reply:
> Ah, I see. Then, let's calculate per-branch tier scores here and aggregate them later? So each branch will have:

{
  tier1: <score, passing>,
  tier2: <score, passing>,
 ...
}
which will be later aggregated across all branches. Would that work?

I think this would work. However, it goes against the idea of separating the check from score calculation because we now would be doing score calculation at the same time that we retrieve settings. The current code just retrieves data, and score calculation is done after, which is what we want in the long term. Btw, we also need the max score for each branch for sanity checking, I think.

Reply:
Sg, lgtm